zipline/pipeline/loaders/quarter_estimates.py

```diff
@@ -0,0 +1,391 @@
+    QuarterEstimatesLoader to serve the BoundColumns described by
+    `columns`.
+    """
+    required = required_estimates_fields(columns)
+    received = set(events.columns)
+    missing = required - received
+    if missing:
+        raise ValueError(
+            "QuarterEstimatesLoader missing required columns {missing}.\n"
+            "Got Columns: {received}\n"
+            "Expected Columns: {required}".format(
+                missing=sorted(missing),
+                received=sorted(received),
+                required=sorted(required),
+            )
+        )
+
+
+class QuarterEstimatesLoader(PipelineLoader):
+    def __init__(self,
+                 estimates,
+                 base_column_name_map):
+        validate_column_specs(
+            estimates,
+            base_column_name_map
+        )
+
+        self.estimates = estimates[
+            estimates[EVENT_DATE_FIELD_NAME].notnull() &
+            estimates[FISCAL_QUARTER_FIELD_NAME].notnull() &
+            estimates[FISCAL_YEAR_FIELD_NAME].notnull()
+        ]
+
+        self.base_column_name_map = base_column_name_map
+
+    @abstractmethod
+    def load_quarters(self, num_quarters, last, dates):
+        pass
+
+    def get_requested_data_for_col(self, stacked_last_per_qtr, idx, dates):
+        """
+        Selects the requested data for each date.
+
+        Parameters
+        ----------
+        stacked_last_per_qtr : pd.DataFrame
+            The latest estimate known per sid per date per quarter with the
+            dates, normalized quarter, and sid as the index.
+        idx : pd.MultiIndex
+            The index of the row of the requested quarter from each date for
+            each sid.
+        dates : pd.DatetimeIndex
+            The calendar dates for which estimates data is requested.
+
+        Returns
+        --------
+        requested_qtr_data : pd.DataFrame
+            The DataFrame with final values for the requested quarter for all
+            columns; `dates` are the index and columns are a MultiIndex with
+            sids at the top level and the dataset columns on the bottom.
+        """
+        requested_qtr_data = stacked_last_per_qtr.loc[idx]
+        # We no longer need this in the index, but we do need it as a column
+        # to calculate adjustments.
+        requested_qtr_data = requested_qtr_data.reset_index(
+            SHIFTED_NORMALIZED_QTRS
+        )
+        (requested_qtr_data[FISCAL_YEAR_FIELD_NAME],
+         requested_qtr_data[FISCAL_QUARTER_FIELD_NAME]) = \
+            split_normalized_quarters(
+                requested_qtr_data[SHIFTED_NORMALIZED_QTRS]
+            )
+        # Move sids into the columns. Once we're left with just dates
+        # as the index, we can reindex by all dates so that we have a
+        # value for each calendar date.
+        requested_qtr_data = requested_qtr_data.unstack(
+            SID_FIELD_NAME
+        ).reindex(dates)
+        return requested_qtr_data
```

Comment:
The starting index is built from `dates` and from that point dates are only removed; why would it not be a subset?